logstash-core/src/main/java/org/logstash/ackedqueue/Page.java

```diff
@@ -164,6 +164,9 @@ public boolean ack(long firstSeqNum, int count, int checkpointMaxAcks) throws IO
             if (this.writable) {
                 headPageCheckpoint();
             } else {
+                // keep checkpoint up-to-date before purge. In case crash happens in the middle of purging,
```

Comment:
This block of code, now, becomes quite familiar to `checkpoint()` method, WDYT to rewrite this as:
```java
checkpoint();

if (!this.writable) {
    purge();
    final CheckpointIO cpIO = queue.getCheckpointIO();
    cpIO.purge(cpIO.tailFileName(pageNum));
}
```